### ISSUE TITLE
Crawler and node tracker using reth p2p

### ISSUE THREAD

**Issue by Rjected** [paradigmxyz/reth#4778]
## Motivation

[Ethernodes](https://ethernodes.org) is a public-facing node tracker used for ethereum nodes, and even serves as [one of the data sources for execution client diversity](https://clientdiversity.org/methodology/). It has information such as:
 * Client version (user agent, e.g. `geth/v1.11.6-stable/linux-amd64/go1.20.3`)
 * Best block
 * Current difficulty
 * Time since "last seen"

It's definitely an interesting data source, but because it is not open source, it's hard to know the exact methodology used to produce the data. Given the metadata it has about nodes, it is probably using a p2p crawler underneath, and at least gaining information from the [`eth` Status handshake message](https://github.com/ethereum/devp2p/blob/master/caps/eth.md#status-0x00).

There _is_ a node crawler embedded in geth, [`devp2p crawl`](https://github.com/ethereum/go-ethereum/blob/master/cmd/devp2p/crawl.go). `devp2p crawl` is useful as a simple reference implementation of an ethereum crawler.

It would be interesting to build a crawler that is:
 * at least as good as `devp2p crawl`
 * open source, making the methodology more transparent than ethernodes

It would then be very easy to do research on the p2p network in rust, and better understand node usage trends.

## How ethereum node crawlers work

Crawlers do a few main things:
 * Crawl a node discovery network
   * In ethereum: discv4, dnsdisc (implemented in reth), discv5 (not implemented in reth yet)
 * Establish sessions with discovered nodes
 * Perform some sort of validation on the discovered nodes
   * Lots of data can be easily faked or sybil-ed, and connections are easy to spin up! This is mostly an inherent property of permissionless p2p networks, like DHTs.
 * Persist metadata obtained during the above steps
 
The first step is to crawl a discovery network to obtain Peer IDs and IPs to establish connections with. Ethereum can be challenging in that discv4 contains nodes from not only ethereum, but other networks as well (BSC, testnets, etc.). Reth already implements two methods to mitigate this concern:
 * [dnsdisc, or EIP-1459](https://eips.ethereum.org/EIPS/eip-1459), which uses public endpoints [populated with network-curated peers](https://github.com/ethereum/discv4-dns-lists)
 * [EIP-2124](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-2124.md), which adds the node's fork ID to discv4 records, making it possible to filter out nodes from different networks before establishing a connection. Trying to establish connections with nodes that don't share a network wastes valuable time that can be spent crawling instead.

There are three steps to establishing connections once a peer is found:

[![](https://mermaid.ink/img/pako:eNqNkc1OwzAQhF_F2nNaJW7jpj5U4qcSSIAQuSEf6sZLYzWxg-2glqrvjpuWE0iwJ2t2vl1r9gCVVQgcPL73aCq81XLjZCsMidVJF3SlO2kCuWp0hT_la7s-i_ExGi0Wg42Tl4fnHVne3C9LUkujfC23F3gwnJwR4OQ3dtXRbkXusGksadF7ufk_iqFekTLI0Pu_2ScbkNgPdOduMrTKSGlryNIHuW60r1EJIwwk0KJrpVYxrMOJFxBqbFFAHAdKuq0AYY7RJ_tgy72pgAfXYwJ9p2T4Dhb4m2x8VGN6wA-wAz7Px8U0zYtJkWaUzfNJAnvgGaXjgrJpxtIsZzQ_JvBpbeTT8Ww-oZTRYkYZm-a0SACVDtY9nk85XHRY8DoAp18cvwAp55d4?type=png)](https://mermaid.live/edit#pako:eNqNkc1OwzAQhF_F2nNaJW7jpj5U4qcSSIAQuSEf6sZLYzWxg-2glqrvjpuWE0iwJ2t2vl1r9gCVVQgcPL73aCq81XLjZCsMidVJF3SlO2kCuWp0hT_la7s-i_ExGi0Wg42Tl4fnHVne3C9LUkujfC23F3gwnJwR4OQ3dtXRbkXusGksadF7ufk_iqFekTLI0Pu_2ScbkNgPdOduMrTKSGlryNIHuW60r1EJIwwk0KJrpVYxrMOJFxBqbFFAHAdKuq0AYY7RJ_tgy72pgAfXYwJ9p2T4Dhb4m2x8VGN6wA-wAz7Px8U0zYtJkWaUzfNJAnvgGaXjgrJpxtIsZzQ_JvBpbeTT8Ww-oZTRYkYZm-a0SACVDtY9nk85XHRY8DoAp18cvwAp55d4)

> [!NOTE]  
> Everything up to this point has already been implemented in reth, and serves as an introduction to ethereum networking.
> The tools to discover peers, filter those peers by network, and establish connections, can be accomplished by using existing reth networking libraries.

Once a crawler successfully establishes a connection with another node, it can perform extra validation specific to the protocol, to make it more difficult to sybil the crawler. On ethereum, this might be something like sampling some random set of block hashes. No method makes it entirely impossible to sybil the crawler, but it's worth thinking about simple strategies that can distinguish a "real" node from a fake peer on the network.

The metadata produced from the handshake can then be tracked and eventually persisted:
* Timestamps
* IPs
* The `Hello` handshake message
* The remote public key / peer ID
* The `Status` handshake message
* Generally, the transcript of the entire session.
* maybe more??

## Basic crawler MVP

Ultimately the crawler is supposed to produce a list of peers that could be considered "real" based on some set of heuristics, and allow access to the metadata in a useful way. After obtaining a stream or list of peers, the open problems are:
* How should this data be structured?
  * Should information be indexed somehow? For example metadata could be indexed by peer ID, or by IP.
* Where should the data be stored?
  * A prototype should probably store everything in-memory, but `devp2p crawl` for example saves to a JSON file. It's also worth thinking about how this could use a database.
* How should users access this data?
  * Will users primarily access the data once it's persisted, like with `devp2p crawl`, or should there be a network API?
  * If there is an API, what will it look like? What endpoints are the most important or useful?

### Timeseries tracking

It should be possible to specify a crawling interval, so network data can be stored as a timeseries. [Bitnodes](https://bitnodes.io/dashboard/1y/) for example, is able to track user agents / client versions over time:
<img width="1205" alt="bitnodes" src="https://github.com/paradigmxyz/reth/assets/6798349/ce200498-91e7-4ece-95ba-ce53917a16b5">

This would be useful for tracking whether or not users are upgrading to new versions of reth, and identify trends in overall reth usage. **This feature would also mean the crawler surpasses all existing node trackers in functionality.**

### Examples of p2p code in reth
 * https://github.com/paradigmxyz/reth/pull/4736
 * [How to use the networks as a standalone component](https://github.com/paradigmxyz/reth/blob/main/examples/network.rs)
 * [`reth p2p`](https://github.com/paradigmxyz/reth/blob/main/bin/reth/src/p2p/mod.rs)

This article is also a great introduction to the reth p2p stack: [Diving into the Reth p2p stack](https://fiber.chainbound.io/blog/reth-p2p/)

### Other node trackers
* [etherscan](https://etherscan.io/nodetracker)
* [execution-diversity.info](https://execution-diversity.info/) - self reported, see [methodology section of clientdiversity.org](https://clientdiversity.org/methodology/)

**Comment by chirag-bgh** [comment]
Hey @Rjected , 
I and @alessandromazza98 are interested to take this.

**Comment by chirag-bgh** [comment]
> hey @chirag-bgh/ @alessandromazza98 - I started digging into the design for this a bit earlier as well. Happy to help you both contribute, maybe we can start a TG groupchat?

sounds good! let me set up the tg

**Comment by gakonst** [comment]
RE: DB choice I would not overthink it, I would primary key users by IP (or other identifier), and put each field as a column, probably on a Postgres, and call it a day? For aggregate queries you can just query the SQL, the rows will be few thousands/tens of thousands so it'll be instantly queriable, and you get all the nice auto API generators on top of postgres if needed

**Comment by 0xprames** [comment]
Yeah makes sense - that definitely will work. I want to make sure we can support timeseries data charting with pg, but otherwise just keeping it simple DB wise is fine.

I've shared the draft design doc on our TG with the HLD, and a few low level design decisions as well. I think should be good to start implementing the service. 

Open to sharing a copy here as well - i could merge it into the repo as an MD and share a link.

Working on refactoring the current crawler code now (rn the codebase we have just subscribes to disc updates and persists to a file, but we aren't "crawling' like geth does), and then going to code up out the apiserver/db wrapper

**Comment by 0xprames** [comment]
updating here to make sure this issue doesn't go stale - we've got :
<img width="814" alt="Screenshot 2023-10-18 at 4 42 16 PM" src="https://github.com/paradigmxyz/reth/assets/134806363/2bf0f9a3-f1eb-4bcd-b844-64935f3d9d5f">

~10k unique peers from less than a day of running our crawler (forcing "rotation" via restarts) - this was run ~4-5 days ago

the apiserver also seems to work from local testing cc @alessandromazza98 who has been working on that! 

Hoping to ship this soon (possibly by EOW, 2 weeks max) - currently working on ensuring crawler works without any slowdowns/perf issues (and matches the crawler impl here https://github.com/ethereum/node-crawler/)

**Comment by gakonst** [comment]
Excellent! We would be happy to host this under diversity.paradigm.xyz or something similar. Whether it's a raw dataset that we semi-regularly update, or a proper dashboard.

**Comment by 0xprames** [comment]
hey friends

rethcrawler.xyz is the current (mvp) launch!

probably needs a few things - but i think we achieved the spirit of the goal from the initial issue. 

Feel free to work with @alessandromazza98 on anything else (domain migration, code cleanup etc)

there has been some discussion on extending this to track CL stats, and i think its a good idea. but maybe out of scope for this specific issue.

there is the outstanding question of infra, we definitely want to migrate off of aws  - it was relatively quick to setup for this initial mvp, but it's a bit pricey.

I think @0xsmit had a good point w/ digital ocean. I've opened https://github.com/Keep-Reth-Strange/reth-crawler/issues/70 to track this, and as discussed he may be taking this forward. it should be easy enough to replicate this setup with a cheaper infra provider!
